nucypher/blockchain/eth/agents.py

```diff
@@ -38,25 +38,29 @@
     STAKING_INTERFACE_CONTRACT_NAME,
     STAKING_INTERFACE_ROUTER_CONTRACT_NAME
 )
-from nucypher.blockchain.eth.decorators import validate_checksum_address
+from nucypher.blockchain.eth.decorators import ContractInterfaces, contract_api, validate_checksum_address
 from nucypher.blockchain.eth.events import ContractEvents
 from nucypher.blockchain.eth.interfaces import BlockchainInterfaceFactory
 from nucypher.blockchain.eth.registry import AllocationRegistry, BaseContractRegistry
 from nucypher.blockchain.eth.utils import epoch_to_period
 from nucypher.crypto.api import sha256_digest
 
 
+#  TODO: Agents type alias
+# Agents = Union['AdjudicatorAgent', '']
+
+
 class ContractAgency:
     # TODO: Enforce singleton - #1506 - Okay, actually, make this into a module
 
     __agents = dict()
+    TRANSACTION_REGISTRY = dict()
 
     @classmethod
     def get_agent(cls,
-                  agent_class,
+                  agent_class: Type['EthereumContractAgent'],
                   registry: BaseContractRegistry = None,
-                  provider_uri: str = None,
-                  ) -> 'EthereumContractAgent':
+                  provider_uri: str = None):
```

Comment:
If the returned object is of type `agent_class` (that is, there's no funny business in the `agent_class` constructor), the return annotation can be preserved here by defining something like `AgentClass = TypeVar('AgentClass', EthereumContractAgent)` and using it for `agent_class` and return annotation.

Reply:
Oh that's lovely :+1:

Reply:
On second look - the return type is actually an instance of a subclass of `EthereumContractAgent`.  Does that mean we need to use a `Union` of all agent subclasses?

Reply:
Yep, that's what the typevar is for. I just checked it, seems to work as expected, except I forgot that you need to use `bound`:

```
from typing import TypeVar, Type

class A: pass
class B(A): pass
class C(A): pass

T_A = TypeVar('T_A', bound=A)

def test(tp:Type[T_A]) -> T_A:
    return tp()

a = test(A)
reveal_type(a) # should be A
b = test(B)
reveal_type(b) # should be B
c = test(C)
reveal_type(c) # should be C
```

Reply:
:heart_eyes: :heavy_check_mark: